Add FormData.constructor submitter tests

jenseng opened the pull request:
Add 7 new tests around the `FormData.constructor` `submitter` argument (new feature proposed [here](https://github.com/whatwg/xhr/pull/366)). This is ready for review in the context of that proposal, but shouldn't be merged until it is accepted.
* For implementations of the existing spec, it's expected that 5 of these will fail and 2 will succeed (`FormData construction should allow a null submitter` and `The constructed FormData object should not contain an entry for an unnamed submit button submitter`), as can be seen [here](https://github.com/web-platform-tests/wpt/pull/37895/checks?check_run_id=10908719479).
* For implementations of the proposal (e.g. [here](https://github.com/jsdom/jsdom/pull/3496)), they should all pass.

jenseng committed:
Add FormData.constructor submitter tests

jenseng commented:
I just realized there are also a bunch of `FormData` constructor tests over in `xhr/formdata`, should I move these new tests over there instead?

jenseng commented on html/semantics/forms/form-submission-0/constructing-form-data-set.html:
```diff
@@ -158,4 +174,48 @@
+  let entries = [...formData.entries()];
+
+  assert_deep_equals(entries, [["n1", "v1"], ["n3", "v3"]]);
+}, 'The constructed FormData object should not contain entries for an unactivated Image Button submitter');
```
I believe this test is correct based on my reading of [the spec](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#constructing-the-form-data-set), i.e. `before this algorithm was invoked the user indicated a coordinate` would imply that no entries should be added in the negative case. But it could also be argued that that is undefined behavior, since it doesn't explicitly `continue` in the prose.

In looking at implementations of the algorithm, it seems that Blink and WebKit both short circuit if the control wasn't activated, while Gecko will default the coordinate to 0:0.

Would y'all recommend leaving this test in (and getting that fixed in Gecko), or removing the test and considering it undefined behavior, since the spec is slightly unclear?

As far as fixing Gecko to check the activated-ness, my hunch is that should be pretty safe/straightforward, since in normal submissions the submitter would have activated. It's only with the new FormData submitter parameter that there might be an unactivated submitter.

annevk replied:
I think 3 makes the most sense here. Thanks for catching that!

cc @domenic @tkent-google

annevk replied:
https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cform%20action%3D%2Futilities%2Fcgi%2Ftest-tools%2Fecho%3E%0A%20%3Cinput%20type%3Dimage%3E%0A%3C%2Fform%3E%0A%3Cscript%3E%0Aonload%20%3D%20()%20%3D%3E%20document.querySelector(%22input%22).click()%0A%3C%2Fscript%3E also shows this is a clear bug in HTML. I'm surprised we didn't catch this before.

annevk replied:
So @jenseng as for next steps:

1. Leave the test in, but change the pass condition.
2. File an issue against or directly PR HTML.

Given my test above this is a (long overdue given how fundamental the algorithm is) bug fix.

jenseng replied:
@annevk sounds good, I'll tweak the test accordingly and file a separate issue/PR.

One slight clarification though, does your test actually demonstrate a bug in the spec? The `click()` method [should trigger the activation behavior](https://html.spec.whatwg.org/#activation), and when an Image Button is activated it [sets the coordinate to `(0,0)`](https://html.spec.whatwg.org/#image-button-state-(type=image):input-activation-behavior) if the user didn't explicitly pick one.

I think the ambiguity/discrepancy is just for Image Button submitters that were not explicitly activated (e.g. when using `aForm.requestSubmit(aSubmitter)` or the proposed `new FormData(aForm, aSubmitter)`. Let me know If I'm missing something.

annevk replied:
Oh lovely, more action-at-a-distance. You're quite right.

https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cform%20action%3D%2Futilities%2Fcgi%2Ftest-tools%2Fecho%3E%0A%20%3Cinput%20type%3Dimage%3E%0A%3C%2Fform%3E%0A%3Cscript%3E%0Aonload%20%3D%20()%20%3D%3E%20document.querySelector(%22form%22).requestSubmit(document.querySelector(%22input%22))%0A%3C%2Fscript%3E (using `requestSubmit()`) demonstrates it better.

I guess once we have the inline defaulting we should simplify that activation behavior language at the same time.

jenseng commented:
Seems like the build is failing due to some webdriver-bidi export ambiguity around `events`, though I'm not sure why it didn't fail before 🤔 ... looks similar to these recent issues (https://github.com/w3c/webdriver-bidi/pull/348, https://github.com/w3c/webdriver-bidi/pull/349). Should I just do some `link-defaults` stuff in the current PR, or would it be better to get webdriver-bidi fixed?

jenseng commented:
Ok, made the `unactivated Image Button submitter` test expect a `(0,0)` coordinate, and also added a test for an explicit null submitter.

Also refactored/moved the new tests over into `xhr/formdata`, as that feels like a better spot for them.

All tests pass on my Blink/Gecko/WebKit POCs 🥳
